bundles/framework/layerlist/view/LayerCollapse/StateHandler.js

```diff
@@ -0,0 +1,106 @@
+export class StateHandler {
+    constructor () {
+        this.listeners = [];
+        this.selectedLayerIds = [];
+        this.groups = [];
+        this.filterKeyword = null;
+        this.openGroupTitles = null;
+        this.filtered = null;
+        this.map = Oskari.getSandbox().getMap();
+    }
+    getState () {
+        return {
+            mutator: this.getMutator(),
+            selectedLayerIds: this._getSelectedLayerIds(),
+            groups: this.groups,
+            filtered: this.filtered,
+            openGroupTitles: this.openGroupTitles,
+            mapSrsName: this.map.getSrsName()
+        };
+    }
+    updateStateWithProps ({groups, selectedLayerIds, filterKeyword}) {
+        this.groups = groups || this.groups;
+        this.selectedLayerIds = selectedLayerIds || this.selectedLayerIds;
+        if (!this._filterStateChanged(filterKeyword)) {
+            return;
+        }
+        this.filterKeyword = filterKeyword;
+        this.filtered = this._getSearchResults();
+        this.openGroupTitles = this.filtered
+            ? this.filtered.map(result => result.group.getTitle()) : [];
+    }
+    updateSelectedLayerIds () {
+        this.selectedLayerIds = this.map.getLayers().map(layer => layer.getId());
+    }
+    _filterStateChanged (nextFilterKeyword) {
+        return this.filterKeyword !== nextFilterKeyword;
+    }
+    _getSelectedLayerIds () {
+        if (this.selectedLayerIds.length === 0) {
+            this.updateSelectedLayerIds();
+        }
+        return this.selectedLayerIds;
+    }
+    _getSearchResults () {
+        if (!this.filterKeyword && this.filterKeyword !== 0) {
+            return null;
+        }
+        const results = this.groups.map(group => {
+            const layers = group.getLayers()
+                .filter(lyr => group.matchesKeyword(lyr.getId(), this.filterKeyword));
+                // and some other rule?
+            return {group, layers};
+        }).filter(result => result.layers.length !== 0);
+        return results;
+    }
+    addListener (listener) {
+        this.listeners.push(listener);
+    }
+    notify () {
+        this.listeners.forEach(consumer => consumer());
```

Comment:
We should give the current state of the statehandler as argument to consumers, so that they can use it to `setState()` or ReactDOM.render();